### ISSUE TITLE
Stuck while do path completion with too many files

### ISSUE THREAD

**Issue by mitnk** [murarth/linefeed#50]
```
$ cd linefeed/
$ mkdir tmp
$ python
>>> for i in range(1000):
>>>     open('./tmp/f-{}.txt'.format(i), 'wb').close()
>>>
>>> exit()
```

now we have 1000 files under `./tmp/`

```
$ cargo run --example path_completion
path> ./tmp/f-<TAB><TAB>
```

Now we stuck in path completing.

**Comment by mitnk** [comment]
had a quick check:

1) `Display all {} possibilities` does not get showed up, need a `flush`?
2) `self.read.try_read_char(None)` alway got a `None` (when hitting `n` or `y`), thus enter a dead loop.

**Comment by murarth** [comment]
1. Yes, indeed. There should be a `flush` call. Maybe in a few places.
2. Ugh. This is actually a bigger problem than it might seem. When asynchronous functionality was introduced, `try_read_char` was changed to only take input from internal buffers -- because only `read_line_step` can perform a terminal read (and only once per call).
I didn't realize at the time that it would break probably every usage of `try_read_char`.

I'll need to give some thought to how to solve this issue effectively.

Thanks for catching this issue.

**Comment by murarth** [comment]
I've pushed a commit to fix this issue.

It removes `try_read_char` and introduces a state machine to handle input for commands that read input, including completion paging.

**Comment by mitnk** [comment]
Thanks for  this huge yet awesome commit! Now it works as expected to me.

Would you like to release a new version soon for this :)

**Comment by murarth** [comment]
Soon, yes. I'd like to wait a few days to see if any bugs shake out.

**Comment by murarth** [comment]
@mitnk: I've just released version 0.5.2 containing this fix.

**Comment by mitnk** [comment]
Thank you @murarth, great work!
